service/core/kafka/server/src/main/scala/com/lightbend/lagom/internal/broker/kafka/TopicProducerActor.scala

```diff
@@ -226,37 +221,49 @@ private[lagom] class TopicProducerActor[Message](
     brokerList.zip(daoFuture)
   }
 
-  private def eventsPublisherFlow(endpoints: String, offsetDao: OffsetDao) =
+  private def eventsPublisherFlow(
+      endpoints: String,
+      offsetDao: OffsetDao
+  ): Flow[InternalTopicProducerCommand[Message], Future[AkkaOffset], Any] =
     Flow.fromGraph(GraphDSL.create(kafkaFlowPublisher(endpoints)) { implicit builder => publishFlow =>
       import GraphDSL.Implicits._
-      val unzip = builder.add(Unzip[Message, AkkaOffset])
-      val zip   = builder.add(Zip[Any, AkkaOffset])
-      val offsetCommitter = builder.add(Flow.fromFunction[(Any, AkkaOffset), Future[AkkaOffset]] {
-        case (_, akkaOffset) =>
-          offsetDao.saveOffset(akkaOffset).map(done => akkaOffset)
+
+      val offsetCommitter = builder.add(Flow.fromFunction {
+        results: ProducerMessage.Results[String, Message, AkkaOffset] =>
+          val offset = results.passThrough
+          offsetDao.saveOffset(offset).map(_ => offset)
       })
 
-      unzip.out0 ~> publishFlow ~> zip.in0
-      unzip.out1 ~> zip.in1
-      zip.out ~> offsetCommitter.in
-      FlowShape(unzip.in, offsetCommitter.out)
+      publishFlow.out ~> offsetCommitter.in
+
+      FlowShape(publishFlow.in, offsetCommitter.out)
     })
 
-  private def kafkaFlowPublisher(endpoints: String): Flow[Message, _, _] = {
+  private def kafkaFlowPublisher(
+      endpoints: String
+  ): Flow[InternalTopicProducerCommand[Message], ProducerMessage.Results[String, Message, AkkaOffset], _] = {
     def keyOf(message: Message): String = {
       partitionKeyStrategy match {
         case Some(strategy) => strategy(message)
         case None           => null
       }
     }
 
-    Flow[Message]
-      .map { message =>
-        ProducerMessage.Message(new ProducerRecord[String, Message](topicId, keyOf(message), message), NotUsed)
-      }
-      .via {
-        ReactiveProducer.flexiFlow(producerSettings(endpoints))
+    def toRecord(message: Message): ProducerRecord[String, Message] =
+      new ProducerRecord[String, Message](topicId, keyOf(message), message)
+
+    Flow[InternalTopicProducerCommand[Message]]
+      .map[ProducerMessage.Envelope[String, Message, AkkaOffset]] {
+        case InternalTopicProducerCommand.EmitMultipleAndCommit(messages, offset) =>
+          ProducerMessage.MultiMessage(messages.map(toRecord), offset)
```

Comment:
Hmm, the implementation looks trivial. I'm not familiar with the details of MultiMessage, though.

Reply:
reverted